pkg/genlib/generator_with_template.go

```diff
@@ -0,0 +1,529 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package genlib
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Pallinder/go-randomdata"
+	"github.com/lithammer/shortuuid/v3"
+)
+
+var templateFieldMap map[string][]byte
+var trailingTemplate []byte
+
+// GeneratorWithTemplate is resolved at construction to a slice of emit functions
+type GeneratorWithTemplate struct {
+	emitFuncs []emitF
+}
+
+func generateTemplateFromField(cfg Config, fields Fields) []byte {
+	templateBuffer := bytes.NewBufferString("{")
+	for i, field := range fields {
+		fieldWrap := fieldValueWrapByType(field)
+		if fieldCfg, ok := cfg.GetField(field.Name); ok {
+			if fieldCfg.Value != nil {
+				fieldWrap = ""
+			}
+		}
+
+		fieldTrailer := []byte(",")
+		if i == len(fields)-1 {
+			fieldTrailer = []byte("}")
+		}
+
+		if strings.HasSuffix(field.Name, ".*") || field.Type == FieldTypeObject || field.Type == FieldTypeNested || field.Type == FieldTypeFlattened {
+			// This is a special case.  We are randomly generating keys on the fly
+			// Will set the json field name as "field.Name.N"
+			N := 5
+			for i := 0; i < N; i++ {
+				if string(fieldTrailer) == "}" && i < N-1 {
+					fieldTrailer = []byte(",")
+				}
+
+				fieldNameRoot := replacer.Replace(field.Name)
+				fieldTemplate := fmt.Sprintf(`"%s.%d": %s{{.%s.%d}}%s%s`, fieldNameRoot, i, fieldWrap, fieldNameRoot, i, fieldWrap, fieldTrailer)
+				templateBuffer.WriteString(fieldTemplate)
+			}
+		} else {
+			fieldTemplate := fmt.Sprintf(`"%s": %s{{.%s}}%s%s`, field.Name, fieldWrap, field.Name, fieldWrap, fieldTrailer)
+			templateBuffer.WriteString(fieldTemplate)
+		}
+	}
+
+	return templateBuffer.Bytes()
+}
+
+func NewGeneratorWithTemplate(template []byte, cfg Config, fields Fields) (*GeneratorWithTemplate, error) {
+	if len(template) == 0 {
+		template = generateTemplateFromField(cfg, fields)
+	}
+
+	tokenizer := regexp.MustCompile(`([^{]*)({{\.[^}]+}})*`)
```

Comment:
the logic here is to split on `here {{.what.ever0}} there {{.what.ever1}} after`
and have a map of `what.everX` with what must be prefixed (`here `, ` there `) to the random generator value for `what.everX`

this way we can keep writing to the buffer sequentially (prefix and random value), without replace or real template usage: this seems to be the most performant solution so far

replace strategy:
```
Benchmark_Generator
Benchmark_Generator-16                	   13316	     89499 ns/op
Benchmark_GeneratorWithTemplate
Benchmark_GeneratorWithTemplate-16    	      92	  13622401 ns/op
```

write to buffer prefix and random value:
```
Benchmark_Generator
Benchmark_Generator-16                	    9391	    127684 ns/op
Benchmark_GeneratorWithTemplate
Benchmark_GeneratorWithTemplate-16    	    4687	    260068 ns/op
```

we write finally any trailing suffix that has no `.what.everX` following (ie: ` after` in our example)

the logic about `fieldPrefixBuffer` has to be validated by unit tests: this logic should be extracted to a separate function for this purpose

Reply:
final benchmark:
```
Benchmark_Generator-16                     19896             60363 ns/op            3799 B/op         22 allocs/op
Benchmark_GeneratorWithTemplate-16         28335             42998 ns/op             482 B/op         15 allocs/op
```

the difference is from getting rid of [bytes.ReplaceAll](https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/36/commits/87523f82cc5ce235dcebcf93a2463c8ce2cc5718#diff-9e27cdd9458e6fafda86103e129642da8680dd2c3cd87947bb5415f5045bd689L500) in `makeDynamicStubWithTemplate`.

this changes a little the behaviour from legacy generator:
in legacy generator `fields.*` generates from 0 to 5 `fields.keyXYZ`, different for every events. template one generates from 0 to 5 `fields.keyXYZ` as "schema" and they will be the same (with different values) across all events

(indeed the latter is more realistic, at least for the AWS case, where stuff like "dimensions", "labels" and "tags" are usually more the same than less, across a single AWS resource type. things might be different for pods labels maybe, in k8s)